text/0098-store-multiple-replay-segments-in-a-single-blob.md

```diff
@@ -0,0 +1,91 @@
+- Start Date: 2023-05-24
+- RFC Type: decision
+- RFC PR: https://github.com/getsentry/rfcs/pull/98
+- RFC Status: draft
+
+# Summary
+
+Recording data is sent in segments. Each segment is written to its own file. Writing files is the most expensive component of our Google Cloud Storage usage. It is also the most expensive component, in terms of time, in our processing pipeline. By merging many segment files together into a single file we can minimize our costs and maximize our processing pipeline's throughput.
+
+# Motivation
+
+1. Minimize costs.
+2. Improve throughput.
+3. Enable new features in a cost-effective manner.
+
+# Background
+
+This document exists to inform all relevant stakeholders of our proposal and seek feedback prior to implementation.
+
+# Supporting Data
+
+Google Cloud Storage lists the costs for writing and storing data as two separate categories. Writing a file costs $0.005 per 1000 files. Storing that file costs $0.02 per gigabyte. For the average file this works out to: $0.000000012 for storage and $0.00000005 for the write.
+
+In practical terms, this means 75% of our spend is allocated to writing new files.
+
+# Proposal
+
+First, a new table called "recording_byte_range" with the following structure is created:
+
+| replay_id | segment_id | path     | start | stop  |
+| --------- | ---------- | -------- | ----- | ----- |
+| A         | 0          | file.txt | 0     | 6242  |
+| B         | 0          | file.txt | 6242  | 8214  |
+| A         | 1          | file.txt | 8214  | 12457 |
+
+Replay and segment ID should be self-explanatory.  Path is the location of the blob in our bucket.  Start and stop are integers which represent the index positions in an exclusive range. This range is a contiguous sequence of related bytes. In other words, the segment's compressed data is contained within the range.
+
+Notice each row in the example above points to the same file but with different start and stop locations. This is implies that multiple segments and replays can be present in  the same file. A single file can be shared by hundreds of different segments.
+
+This table will need to support, at a minimum, one write per segment.
+
+Second, the Session Replay recording consumer will not _commit_ blob data to Google Cloud Storage for each segment. Instead it will buffer many segments and flush them together as a single blob to GCS. Next it will make a bulk insertion into the database for tracking.
```

Comment:
need to think about: possibilities for incidents & data loss in case of a consumer crash or increased load.

would it be better to use redis for the buffer?

what about an SSD based solution so we have some persistence potential? 

I'm also wondering if the flushing to GCS should be in a separate consumer / process altogether